Use babel umd

clayreimann opened the pull request:
~~This should not be merged until Babel `6.6.0` is released.~~ `babel@6.6.x` is released.

This could constitute the beginnings of the rework for the `0.12.x`/`1.0.0` release

To Do:
- [x] Add test coverage back
- ~~Add browser tests with [zuul](https://www.npmjs.com/package/zuul)~~ *Will be fixed by* #300
- ~~Add deploy scripts that generate the `dist/` artifacts~~ *Will be fixed by* #299
- ~~Add auto-generated documentation~~ *Will be fixed by* #239

AurelioDeRosa commented:
@clayreimann the reason why we didn't use Babel anymore is because we'll lose most of the code coverage tooling. While it's definitely possible to have the final metric value (e.g. 85% covered), the covered/uncovered lines will be lost. Also, altought very painful, we still need the dist files as demonstrated by issues like https://github.com/michael/github/issues/285.

clayreimann commented:
I have:

* Moved the testing from `karma+mocha` to just `mocha`
* Fixed the tests to that they will actually run in `mocha` when I checkout from master all I see are timeouts when I try and run the tests
* Moved to using `babel`'s UMD wrapper to clean up our code
* Added a fix for how we handle Blobs in node vs the browser

If this all looks good I'll add one more commit that fixes indentation in `github.js`. I haven't done that yet to keep the diff simple and not muck everything up with whitespace changes.

I think the changes here are going to be critical in allowing us to modularize our codebase going forward.

aendrew commented:
This is a huge amount of work, @clayreimann — thank you very much for it. I've admittedly been a bit distant from how @AurelioDeRosa has been managing the code the last while so it's ultimately up to him, but having gone through most of the PR (albeit fairly quickly), it looks pretty terrific. 

1. It seems all the Sauce Labs stuff has been stripped out. Thank you and good riddance; how I had it set up was woefully broken and any benefit gained from it was outweighed by the additional hassle it caused. It's better to not have it and then set it up later after the codebase has stabilized a bit. So, :+1: from me on that.

2. `npm test` is broken (at least for the version of Gulp I have); either the Gulpfile should be updated so that `gulp test:mocha` is the default task of `test`, or `package.json` updated so that `npm test` starts the test cycle. I'm impartial to either solution.

3. There's still a karma.conf.js even though Karma isn't a dep (unless I'm missing something, it's probably okay to just remove that too). With Karma gone, is there any way (or point) to still test in the browser?

That's all for now, will add more if I see anything else.

clayreimann commented:
@aendrew Thanks for the review. I do plan on helping modularizing the library, but after #264 didn't go anywhere I figured it would be better to do it piecemeal. 

Other changes in a7587fd:
* Fixed `npm test`
* Fixed whitespace linting in `github.js`

AurelioDeRosa commented on .gitignore:
```diff
@@ -1,6 +1,8 @@
-npm-debug.log
 coverage/
+
+.zuulrc
```
What's this file for?

AurelioDeRosa commented on gulpfile.js:
```diff
@@ -1,185 +1,113 @@
+   gulp.start('lint', 'test:mocha', 'build');
 });
-
-var sauceLaunchers = {
```
Does the PR remove sauce lab completely? I remember we discussed  this point, but now I can't remember the reason on top of my mind.

clayreimann replied:
It does remove sauce completely because for PRs Sauce doesn't work. The intent is to add it back for `master` so we can populate the browser matrix.

AurelioDeRosa commented on package.json:
```diff
@@ -1,66 +1,63 @@
+      "jshint": "^2.9.1",
+      "jshint-stylish": "^2.1.0",
+      "mocha": "^2.3.4",
+      "must": "^0.13.1",
```
Why changing of the assertion library and removing karma?

clayreimann replied:
Remove `karma` because we're not doing browser testing any more so `mocha` is sufficient. And changing the assertion library because [both `should` and `mocha` make bad design choices](https://github.com/moll/js-must#asserting-on-property-access)

AurelioDeRosa commented on test/helpers.js:
```diff
@@ -0,0 +1,49 @@
+'use strict';
+
+var expect = require('must');
+var STANDARD_DELAY = 200; // 200ms between nested calls to the API so things settle
```
If we add this delay I feel we aren't testing the way our users will employ the library. Even in my projects I run the next async call as soon as the callback returns. Any comment @aendrew?

clayreimann replied:
The delay allows the Github's filesystem to settle. Before I added the delay I found that the tests that created/updated files/repos failed nondeterministically because Github's system wasn't always fast enough to reflect how we were setting the tests up.

I suppose we could enable it only for writes if you want.

AurelioDeRosa commented on test/helpers.js:
```diff
@@ -0,0 +1,49 @@
+var expect = require('must');
+var STANDARD_DELAY = 200; // 200ms between nested calls to the API so things settle
+
+function assertSuccessful(done, cb) {
```
Is the aim of these helpers to replace only those tests that have those assertions? Stated in other words, if a callback for a test has:

```
expect(err).not.to.exist();
expect(res).to.exist();
expect(xhr).to.be.an.object();
```

I should use this helper, but if the test has more I should not use it?

clayreimann replied:
It's to keep the tests dry. We (almost) always asserted 
```
expect(err).not.to.exist();
expect(res).to.exist();
expect(xhr).to.be.an.object();
```
So I captured that here, along with the common logic for handling errors. If you have more assertions that the three above then you simply pass a callback to this helper and it will be executed as well.

AurelioDeRosa commented on test/test.auth.js:
```diff
@@ -1,59 +1,96 @@
+               expect(xhr).to.be.an.object();
+
+               done();
+            } catch(e) {
```
I'd suggest to avoid names like `e`. They can be confused with `event`.

AurelioDeRosa commented on test/test.auth.js:
```diff
@@ -1,59 +1,96 @@
+                     return;
+                  }
+               } catch(e2) {
+                  done(e);
```
I think this line can be avoided an have an empty catch.

clayreimann replied:
Nope, without `done(e)` the test won't fail if something goes wrong in the preceding `try` block

AurelioDeRosa replied:
You have it already in the next line (65). You capture the error, do nothing, and then execute the `done(e)` in line 65.

AurelioDeRosa commented on test/test.auth.js:
```diff
@@ -1,59 +1,96 @@
-   });
+      // 200ms between tests so that Github has a chance to settle
+      beforeEach(function(done) {
+         setTimeout(done, 200);
```
Same comment about the delay I wrote before.

AurelioDeRosa commented:
In addition to my comments, do you confirm that at this stage the code coverage won't work?

AurelioDeRosa commented:
I'd also say that we probably should start creating a v.1 branch. This way we can start merging all this work plus other things that will come in that branch, but keep merging the PRs other devs are submitting. Once we complete version 1, we can rebase and complete a massive merge.
